docs/design/README.md

```diff
@@ -534,6 +534,149 @@ After installing packages, the result of `dnf list installed` will be output to
+This will be done by traversing the Cargo dependency graph and making a list of packages and kits that are needed.
+The kits will be unpacked and the direct package dependencies will be copied from their `build/rpms` subdirectories.
+This will require us to pass a list of the packages needed for the build to the dockerfile, for example:
+
+```
+PACKAGES=pkg-a pkg-b pkg-c
+```
+
+This list needs to be complete and include only "pure" package dependencies.
+That is, it should not include any package names that are "brought in" by a kit dependency.
+The packages found in each of those subdirectories, e.g. `build/rpms/pkg-a`, will be copied to a flat RPM directory for the build.
+
+Additionally, the `build-package` command needs to know the list of kits required:
+
+```
+KITS=core-kit aws-kit
+```
+
+Each of these needs to be unpacked and have its RPMs copied out to the flat RPM directory.
+For example, these would be unpacked: `build/kits/core-kit.tar`, `build/kits/aws-kit.tar`.
+
+The `build-kit` requirements are slightly simplified.
+Only the list of "pure" package dependencies is required.
+When constructing a kit, buildsys does not need to know what other kits the current kit depends on.
+
+#### Dependency Traversal
+
+The simplest way to get a list of "pure" package dependencies and a list of kits is to traverse the dependency graph twice.
+
+In the first traversal, kit dependencies are ignored and not followed, resulting in only the "pure" package dependencies.
+In the second traversal, all nodes are followed and package dependencies are filtered out from the final list, resulting in a list of all required kits.
+
+#### Pseudo Code
+
+**Package Build Time**
+
+buildsys:
+
+```
+// load the package's manifest
+let (packages, kits) = manifest.walk_dependency_graph();
+
+// when calling build_package:
+  build-arg PACKAGES="{packages}"
+  build-arg KITS="{kits}"
+```
+
+Dockerfile:
+
+```
+ARG PACKAGES
+ARG KITS
+
+mkdir big-rpm-dir
+
+# PACKAGES could be empty for packages like glibc with no deps
+for pkg in ${PACKAGES} ; do
+  cp /host/build/rpms/${pkg}/* big-rpm-dir
+done
+
+# KITS could be empty for core packages with no kits dependency
+for kit in ${KITS} ; do
+  tar xf /host/build/kits/${kit}.tar -C big-rpm-dir
+done
+
+createrepo_c big-rpm-dir
+
+# now run rpmbuild stage using big-rpm-dir as input
+```
+
+**Kit Build Time**
+
+manifest for core kit:
+
+```
+# core kit in kits/core/Cargo.toml
+
+[dependencies]
+glibc = "../../packages/glibc"
+systemd = "../../packages/systemd"
+```
+
+manifest for k8s kit:
+
+```
+# k8s kit in kits/k8s/Cargo.toml
+
+[dependencies]
+kubelet = "../../packages/kubelet"
+
+# FIXME naming convention for these crates?
+core-kit = "../core"
+```
+
+buildsys:
+
+```
+// load the kit's manifest
+let (packages, _) = manifest.walk_dependency_graph();
+
+// no need for kits, they're some other crate's problem
+// when calling build_kit:
+  build-arg PACKAGES="{packages}"
+```
```

Comment:
Just a note talking with Ben the other day we don't necessarily think that buildsys should care about packing the repository into a container. This is still accurate here right, as in the dockerfile is more for orchestrating the kit build for creating the "local-kit"?